posts/2023-12-21-async-fn-rpit-in-traits.md

```diff
@@ -0,0 +1,237 @@
+---
+layout: post
+title: "Announcing `async fn` and return-position `impl Trait` in traits"
+author: Tyler Mandry
+team: The Async Working Group <https://www.rust-lang.org/governance/wgs/wg-async>
+---
+
+The Rust Async Working Group is excited to announce major progress towards our goal of enabling the use of `async fn` in traits. Rust 1.75, which hits stable next week, will include support for both `-> impl Trait` notation and `async fn` in traits.
+
+This is a big milestone, and we know many users will be itching to try these out in their own code. However, we are still missing some important features that many users need. Read on for an recommendations on when and how to use the stabilized features.
+
+## What's stabilizing
+
+Ever since the stabilization of [RFC #1522] in Rust 1.26, Rust has allowed users to write `impl Trait` as the return type of functions (often called "RPIT"). This means that the function returns "some type that implements `Trait`". This is commonly used to return closures, iterators, and other types that are complex or impossible to write explicitly.
+
+[RFC #1522]: https://rust-lang.github.io/rfcs/1522-conservative-impl-trait.html
+
+```rust
+/// Given a list of players, return an iterator
+/// over their names.
+fn player_names(
+    players: &[Player]
+) -> impl Iterator<Item = &String> {
+    players
+        .iter()
+        .map(|p| &p.name)
+}
+```
+
+Starting in Rust 1.75, you can use **return-position `impl Trait` in trait** (RPITIT) definitions and in trait impls. For example, you could use this to write a trait that returns an iterator:
+
+```rust
+trait Container {
+    fn items(&self) -> impl Iterator<Item = Widget>;
+}
+
+impl Container for MyContainer {
+    fn items(&self) -> impl Iterator<Item = Widget> {
+        self.items.iter().cloned()
+    }
+}
+```
+
+So what does all of this have to do with async functions? Well, async functions are "just sugar" for a function that returns `-> impl Future`. Since that is now permitted in traits, **we also permit you to write traits that use `async fn`**.[^desugar-correct]
```

Comment:
```suggestion
So what does all of this have to do with async functions? Well, async functions are "just sugar" for functions that returns `-> impl Future`. Since these are now permitted in traits, **we also permit you to write traits that use `async fn`**.[^desugar-correct]
```